File: scrimage-core/src/main/scala/com/sksamuel/scrimage/canvas/drawable/FilledPolygon.scala
package com.sksamuel.scrimage.canvas.drawable

import java.awt.Graphics2D

import com.sksamuel.scrimage.canvas.{Drawable, GraphicsContext}

case class FilledPolygon(points: Seq[java.awt.Point], context: GraphicsContext) extends Drawable {
  def draw(g: Graphics2D): Unit = {
    g.fillPolygon(points.map(_.x).toArray, points.map(_.y).toArray, points.size)
  }
  def open: Polygon = Polygon(points, context)
}

File: scrimage-core/src/main/java/com/sksamuel/scrimage/canvas/drawables/FilledPolygon.java
package com.sksamuel.scrimage.canvas.drawables;

import com.sksamuel.scrimage.canvas.Drawable;
import com.sksamuel.scrimage.canvas.GraphicsContext;
import com.sksamuel.scrimage.graphics.RichGraphics2D;

import java.util.List;

public class FilledPolygon implements Drawable {

    private final List<java.awt.Point> points;
    private final GraphicsContext context;

    public FilledPolygon(List<java.awt.Point> points, GraphicsContext context) {
        this.points = points;
        this.context = context;
    }

    @Override
    public void draw(RichGraphics2D g) {
        int n = points.size();
        int[] xs = new int[n];
        int[] ys = new int[n];
        int i = 0;
        for (java.awt.Point p : points) {
            xs[i] = p.x;
            ys[i] = p.y;
            i++;
        }
        g.fillPolygon(xs, ys, n);
    }

    public Polygon toOutline() {
        return new Polygon(points, context);
    }

    @Override
    public GraphicsContext context() {
        return context;
    }
}
